File: gke_customization/gke_hrms/report/pf_contribution_report/pf_contribution_report.js
// Copyright (c) 2026, Gurukrupa Export and contributors
// For license information, please see license.txt

frappe.query_reports["PF Contribution Report"] = {
	"filters": [
		{
            "label": __("Month"),
            "fieldtype": "Select",
            "fieldname": "month",
            "reqd": 1,
            "options": [
                "January","February","March","April",
                "May","June","July","August",
                "September","October","November","December"
            ],
            "default": new Date().toLocaleString('en-US', { month: 'long' }),
            "on_change": function() {
                set_dates();
            }
        },
        {
            "label": __("Year"),
            "fieldtype": "Select",
            "fieldname": "year",
            "reqd": 1,
            "options": (() => {
                let years = [];
                let currentYear = new Date().getFullYear();
                for (let i = currentYear - 5; i <= currentYear + 5; i++) {
                    years.push(String(i));
                }
                return years;
            })(),
            "default": String(new Date().getFullYear()),
            "on_change": function() {
                set_dates();
            }
        },
		{
			"label": __("From Date"),
			"fieldtype": "Date",
			"fieldname": "from_date",
			"read_only": 1,
			"default": frappe.datetime.month_start(frappe.datetime.get_today()),
			"on_change": function(query_report) {
				var from_date = query_report.get_values().from_date;
				if (!from_date) {
					return;
				}
				let date = new moment(from_date)
				var to_date = date.endOf('month').format()
				query_report.set_filter_value({
					"to_date": to_date
				});
			}
		},
		{
			"label": __("To Date"),
			"fieldtype": "Date",
			"fieldname": "to_date",
			"reqd": 1,
			"read_only": 1,
			"default": frappe.datetime.month_end(frappe.datetime.get_today())
		},
		{
			"fieldname": "company",
			"label": "Company",
			"fieldtype": "Link",
			"options": "Company",
			// "default": frappe.defaults.get_user_default("Company"),
			"reqd": 1
		},
		//{
		//	"fieldname": "report_type",
		//	"label": "Report Type",
		//	"fieldtype": "Select",
		//	"options": "Register Report\nPF File",
		//	"default": "PF File",
		//	"reqd": 1
		//},
		{
			"fieldname": "employee",
			"label": "Employee",
			"fieldtype": "MultiSelectList",
			"get_data": function(txt) {
				var company = frappe.query_report.get_filter_value('company');
				var filters = {}
				if (company) filters['company'] = company
				return frappe.db.get_link_options('Employee', txt, filters);
			}
		}
	],
	
	onload: (report) => {
		set_dates();
		report.page.add_button("Clear Filters", function() {
			window.open("/app/query-report/PF%20Contribution%20Report", "_self")
		}).addClass("btn-info")	
	}
};

function fetch_month_list() {
	frappe.call({
		method: "gke_customization.gke_hrms.report.pf_contribution_report.pf_contribution_report.get_month_range",
		freeze: false,
		callback: function(r) {
			var month = frappe.query_report.get_filter('month');
				month.df.options.push(...r.message);
				month.refresh();
		}
	})
}

function set_dates() {
    let report = frappe.query_report;
    if (!report) return;

    let monthName = report.get_filter_value("month");
    let year = parseInt(report.get_filter_value("year"));

    if (!monthName || !year) return;

    const monthIndex = [
        "January","February","March","April",
        "May","June","July","August",
        "September","October","November","December"
    ].indexOf(monthName);

    let from_date = new Date(year, monthIndex, 1);
    let to_date = new Date(year, monthIndex + 1, 0);

    report.set_filter_value({
        from_date: frappe.datetime.obj_to_str(from_date),
        to_date: frappe.datetime.obj_to_str(to_date)
    });
}

File: gke_customization/gke_hrms/report/pf_contribution_report/pf_contribution_report.py
# Copyright (c) 2026, Gurukrupa Export and contributors
# For license information, please see license.txt

import frappe
from frappe.utils.dateutils import get_dates_from_timegrain, get_period
from frappe.utils import today, add_to_date

def execute(filters=None):
	columns = get_columns(filters)
	data = get_pf_data(filters)
	return columns, data

def get_pf_data(filters=None):
	# month_name = filters.get("month")
	# year = filters.get("year")

	# month_info_map = {
	# 	"January": ("01", "31"), "February": ("02", "28"), "March": ("03", "31"),
	# 	"April": ("04", "30"), "May": ("05", "31"), "June": ("06", "30"),
	# 	"July": ("07", "31"), "August": ("08", "31"), "September": ("09", "30"),
	# 	"October": ("10", "31"), "November": ("11", "30"), "December": ("12", "31")
	# }

	# month_data = month_info_map.get(month_name)
	# month_num = month_data[0]
	# last_day = month_data[1]

	# year_int = ''
	# if month_name == "February":
	# 	year_int = int(year)
	# 	if (year_int % 4 == 0 and year_int % 100 != 0) or (year_int % 400 == 0):
	# 		last_day = "29"

	# start_date = f"{year}-{month_num}-01"
	# end_date = f"{year}-{month_num}-{last_day}"

	start_date = filters.get("from_date")
	end_date = filters.get("to_date")
 
	emp_condition = ""
	if filters.get("employee"):
		emp_list = filters.get("employee")
		formatted_emps = ", ".join([f"'{e}'" for e in emp_list])
		emp_condition = f"AND ss.employee IN ({formatted_emps})"

	sql_query = f"""
		SELECT 
			ss.employee,
			ss.employee_name,
			ss.department,
			ss.designation,
			ss.total_working_days,
			ss.payment_days,
			ss.actual_working_hours,
			ss.gross_pay,
			ss.total_deduction,
			ss.rounded_total,
			emp.uan_number AS uan, ss.absent_days,
			Least(Ifnull(basic.basic, 0), 15000) AS epf_wage,
			Least(Ifnull(basic.basic, 0), 15000) AS edli_wage,
			CASE WHEN Timestampdiff(year, emp.date_of_birth, Curdate()) < 58 THEN Least(Ifnull(basic.basic, 0), 15000) ELSE 0 end AS eps_wage,
			Round(Least(Ifnull(basic.basic, 0), 15000) * 0.12, 0) AS ee_share,
			CASE WHEN Timestampdiff(year, emp.date_of_birth, Curdate()) < 58 THEN Round(Least(Ifnull(basic.basic, 0), 15000) * 0.0833, 0) ELSE 0 end AS eps_con,
			Round(Least(Ifnull(basic.basic, 0), 15000) * 0.0367, 2) AS er_share,
			CASE
				WHEN IFNULL(ss.absent_days, 0) > 0
					THEN ROUND(ss.absent_days, 0)
				ELSE
					GREATEST(
						ROUND(
							IFNULL(ss.total_working_days, 0)
							- IFNULL(ss.payment_days, 0),
							0
						),
						0
					)
			END AS ncp_days,
			0 as Refund
		FROM `tabSalary Slip` ss
		LEFT JOIN `tabEmployee` emp ON emp.name = ss.employee
		LEFT JOIN (
			SELECT parent, Sum(CASE WHEN salary_component = 'Basic' THEN amount ELSE 0 end) AS basic
			FROM `tabSalary Detail`
			GROUP BY parent
		) basic ON basic.parent = ss.name
		WHERE ss.docstatus = 1
			AND ss.start_date >= '{start_date}'
			AND ss.end_date <= '{end_date}'
			AND ss.company = '{filters.get("company")}'
			AND emp.uan_number IS NOT NULL
			{emp_condition}
		ORDER BY ss.employee_name
		"""
	# frappe.msgprint("<pre>" + sql_query + "</pre>")
	data = frappe.db.sql(sql_query, as_dict=True)
	# frappe.msgprint(f"{data}")
	
	data += get_totals(data, filters.get("employee"), filters)
	return data

@frappe.whitelist()
def get_totals(data, employee=None, filters=None):
	totals = {
		"company": "Total",
		"branch": "",
		"employee_name": "Total Employees: {}".format(len(data)),
		"gross_pay": sum(row.get("gross_pay", 0) for row in data),
		"epf_wage": sum(row.get("epf_wage", 0) for row in data),
		"edli_wage": sum(row.get("edli_wage", 0) for row in data),
		"eps_wage": sum(row.get("eps_wage", 0) for row in data),
		"ee_share": sum(row.get("ee_share", 0) for row in data),
		"eps_con": sum(row.get("eps_con", 0) for row in data),
		"er_share": sum(row.get("er_share", 0) for row in data),
		"ncp_days": sum(row.get("ncp_days", 0) for row in data),
	}
	return [totals]

def get_columns(filters=None):
    columns = [
		{"fieldname": "uan", "label": "UAN", "fieldtype": "Data", "width": 120},
		{"fieldname": "employee_name", "label": "Employee Name", "fieldtype": "Data", "width": 160},
		{"fieldname": "gross_pay", "label": "Gross Pay", "fieldtype": "Currency", "width": 100},
		{"fieldname": "epf_wage", "label": "EPF Wage", "fieldtype": "Currency", "width": 100},
		{"fieldname": "edli_wage", "label": "EDLI Wage", "fieldtype": "Currency", "width": 100},
		{"fieldname": "eps_wage", "label": "EPS Wage", "fieldtype": "Currency", "width": 100},
		{"fieldname": "ee_share", "label": "EE Share", "fieldtype": "Currency", "width": 100},
		{"fieldname": "eps_con", "label": "EPS Contr.", "fieldtype": "Currency", "width": 100},
		{"fieldname": "er_share", "label": "ER Share", "fieldtype": "Currency", "width": 100},
		{"fieldname": "ncp_days", "label": "NCP Days", "fieldtype": "Float", "width": 80},
		{"fieldname": "Refund", "label": "Refund", "fieldtype": "Int", "width": 80}
	]
    
    return columns

@frappe.whitelist()
def get_month_range():
	end = today()
	start = add_to_date(end, months=-12)
	periodic_range = get_dates_from_timegrain(start, end, "Monthly")
	periods = [get_period(row) for row in periodic_range]
	periods.reverse()
	return periods